File: 14-Algorithms_searching+bsf+dfs/depth_first_search.py
class Node:
    def __init__(self, value):
        self.left = None
        self.right = None
        self.value = value


class DepthFirstSearch:
    def __init__(self):
        self.root = None

    def __iter__(self):
        node = self.root
        while node is not None:
            if node.left is not None:
                yield from node.left
                node = node.left

            if node.right is not None:
                yield node.right
                node = node.right

    def insert(self, value):
        new_node = Node(value)

        if self.root is None:
            self.root = new_node
        else:
            current_node = self.root

            while True:
                if value < current_node.value:
                    # Left
                    if current_node.left is None:
                        current_node.left = new_node
                        return

                    current_node = current_node.left

                if value >= current_node.value:
                    # Right
                    if current_node.right is None:
                        current_node.right = new_node
                        return

                    current_node = current_node.right

    def lookup(self, value):
        if self.root is None:
            return False

        current_node = self.root

        while current_node:
            if value < current_node.value:
                current_node = current_node.left
            elif value > current_node.value:
                current_node = current_node.right
            elif value == current_node.value:
                return vars(current_node)

        return False

    def remove(self, value):
        return self.__remove_n(self.root, value)

    def in_order(self):
        return self.__in_order_t(self.root, [])

    def pre_order(self):
        return self.__pre_order_t(self.root, [])

    def post_order(self):
        return self.__post_order_t(self.root, [])

    def __find_min(self, node):
        current_node = node

        while current_node.left:
            current_node = current_node.left

        return current_node

    def __remove_n(self, node, value):
        if node is None:
            return node

        if value < node.value:
            node.left = self.__remove_n(node.left, value)
        elif value > node.value:
            node.right = self.__remove_n(node.right, value)
        else:
            if node.left is None:
                return node.right

            if node.right is None:
                return node.left

            min = self.__find_min(node.right)
            node.value = min.value
            node.right = self.__remove_n(node.right, min.value)

        return node

    def __in_order_t(self, node, res):
        if node:
            self.__in_order_t(node.left, res)
            res.append(node.value)
            self.__in_order_t(node.right, res)

        return res

    def __pre_order_t(self, node, res):
        if node:
            res.append(node.value)
            self.__pre_order_t(node.left, res)
            self.__pre_order_t(node.right, res)

        return res

    def __post_order_t(self, node, res):
        if node:
            self.__post_order_t(node.left, res)
            self.__post_order_t(node.right, res)
            res.append(node.value)

        return res


tree = DepthFirstSearch()
tree.insert(9)
tree.insert(4)
tree.insert(6)
tree.insert(20)
tree.insert(170)
tree.insert(15)
tree.insert(1)
print(tree.lookup(170))
print(tree.in_order())
print(tree.pre_order())
print(tree.post_order())

File: 14-Algorithms_searching+bsf+dfs/depth_first_search.rb
# frozen_string_literal: true

class Node
  attr_accessor :left, :right, :value

  def initialize(value)
    @left = nil
    @right = nil
    @value = value
  end
end

class DepthFirstSearch
  attr_reader :root

  def initialize
    @root = nil
  end

  def insert(value)
    new_node = Node.new(value)

    if @root.nil?
      @root = new_node
    else
      current_node = @root

      while true
        if value < current_node.value
          # Left
          return current_node.left = new_node unless current_node.left

          current_node = current_node.left
        end

        next unless value >= current_node.value
        # Right
        return current_node.right = new_node unless current_node.right

        current_node = current_node.right
      end
    end
  end

  def lookup(value)
    return false if @root.nil?

    current_node = @root

    while current_node
      if value < current_node.value
        current_node = current_node.left
      elsif value > current_node.value
        current_node = current_node.right
      elsif value == current_node.value
        return current_node
      end
    end

    false
  end

  def remove(value)
    remove_n(root, value)
  end

  def in_order
    in_order_t(@root, [])
  end

  def pre_order
    pre_order_t(@root, [])
  end

  def post_order
    post_order_t(@root, [])
  end

  private

  def find_min(node)
    current_node = node
    current_node = current_node.left while current_node.left

    current_node
  end

  def remove_n(node, value)
    return node if node.nil?

    if value < node.value
      node.left = remove_n(node.left, value)
    elsif value > node.value
      node.right = remove_n(node.right, value)
    else
      return node.right if node.left.nil?
      return node.left if node.right.nil?

      min = find_min(node.right)
      node.value = min.value
      node.right = remove_n(node.right, min.value)
    end

    node
  end

  def in_order_t(node, res)
    return unless node

    in_order_t(node.left, res)
    res << node.value
    in_order_t(node.right, res)
    res
  end

  def pre_order_t(node, res)
    return unless node

    res << node.value
    pre_order_t(node.left, res)
    pre_order_t(node.right, res)
    res
  end

  def post_order_t(node, res)
    return unless node

    post_order_t(node.left, res)
    post_order_t(node.right, res)
    res << node.value
    res
  end
end

tree = DepthFirstSearch.new
tree.insert(9)
tree.insert(4)
tree.insert(6)
tree.insert(20)
tree.insert(170)
tree.insert(15)
tree.insert(1)
p tree.lookup(170)
p tree
p tree.in_order
p tree.pre_order
p tree.post_order
